File: typescript/src/domain/payment/PaymentStateMachine.ts
import PaymentState from './PaymentState';
import PaymentEvent from './PaymentEvent';

/**
 * Tabla explícita de transiciones del ciclo de vida del pago.
 *
 * Un estado omitido del mapa (o un evento ausente para un estado dado) implica
 * una transición terminal / ilegal y la máquina la rechaza.
 */
const TRANSITIONS: Partial<Record<PaymentState, Partial<Record<PaymentEvent, PaymentState>>>> = {
    PENDING: {
        AUTHORIZE: 'AUTHORIZED',
        FAIL: 'FAILED',
        CANCEL: 'CANCELLED',
    },
    AUTHORIZED: {
        CAPTURE: 'CAPTURED',
        FAIL: 'FAILED',
        CANCEL: 'CANCELLED',
    },
};

/** Se lanza cuando se despacha un evento que el estado actual no acepta. */
export class IllegalTransitionError extends Error {
    constructor(
        readonly from: PaymentState,
        readonly event: PaymentEvent,
    ) {
        super(`Transición ilegal: no se puede aplicar "${event}" mientras se encuentra en estado "${from}".`);
        this.name = 'IllegalTransitionError';
    }
}

/**
 * Máquina de estados finita determinista para un pago.
 *
 * La máquina solo posee el estado actual y las reglas para cambiarlo; a propósito
 * no incluye persistencia, montos ni efectos secundarios para poder razonar
 * sobre el ciclo de vida de forma aislada.
 */
class PaymentStateMachine {
    private current: PaymentState;

    constructor(initial: PaymentState = 'PENDING') {
        this.current = initial;
    }

    /** Estado en el que se encuentra la máquina actualmente. */
    get state(): PaymentState {
        return this.current;
    }

    /** @returns `true` si el pago no puede transicionar. */
    get isTerminal(): boolean {
        return this.allowedEvents().length === 0;
    }

    /** Eventos aceptados. */
    allowedEvents(): PaymentEvent[] {
        return Object.keys(TRANSITIONS[this.current] ?? {}) as PaymentEvent[];
    }

    /** Indica si `event` sería aceptado desde el estado actual. */
    can(event: PaymentEvent): boolean {
        return TRANSITIONS[this.current]?.[event] !== undefined;
    }

    /** Estado al que llevaría `event`, o undefined si no está permitido. */
    peek(event: PaymentEvent): PaymentState | undefined {
        return TRANSITIONS[this.current]?.[event];
    }

    /**
     * Aplica `event` y avanza al siguiente estado.
     * @throws {IllegalTransitionError} si el evento no está permitido en este estado.
     */
    dispatch(event: PaymentEvent): PaymentState {
        const next = this.peek(event);
        
        if (next === undefined) throw new IllegalTransitionError(this.current, event);
        this.current = next;
        
        return this.current;
    }
}

export default PaymentStateMachine;

File: java/src/main/java/com/jcoppede/fsm/payment/PaymentStateMachine.java
package com.jcoppede.fsm.payment;

import java.util.Collections;
import java.util.EnumMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * Máquina de estados finita determinista para un pago.
 *
 * La máquina solo posee el estado actual y las reglas para cambiarlo.
 * Un estado sin transiciones salientes es terminal.
 */
public class PaymentStateMachine {

    /** Tabla explícita de transiciones: estado -> (evento -> estado destino). */
    private static final Map<PaymentState, Map<PaymentEvent, PaymentState>> TRANSITIONS =
            buildTransitions();

    private static Map<PaymentState, Map<PaymentEvent, PaymentState>> buildTransitions() {
        Map<PaymentState, Map<PaymentEvent, PaymentState>> t = new EnumMap<>(PaymentState.class);

        Map<PaymentEvent, PaymentState> pending = new EnumMap<>(PaymentEvent.class);
        pending.put(PaymentEvent.AUTHORIZE, PaymentState.AUTHORIZED);
        pending.put(PaymentEvent.FAIL, PaymentState.FAILED);
        pending.put(PaymentEvent.CANCEL, PaymentState.CANCELLED);
        t.put(PaymentState.PENDING, pending);

        Map<PaymentEvent, PaymentState> authorized = new EnumMap<>(PaymentEvent.class);
        authorized.put(PaymentEvent.CAPTURE, PaymentState.CAPTURED);
        authorized.put(PaymentEvent.FAIL, PaymentState.FAILED);
        authorized.put(PaymentEvent.CANCEL, PaymentState.CANCELLED);
        t.put(PaymentState.AUTHORIZED, authorized);

        return t;
    }

    private PaymentState current;

    public PaymentStateMachine() {
        this(PaymentState.PENDING);
    }

    public PaymentStateMachine(PaymentState initial) {
        this.current = initial;
    }

    /** Estado en el que se encuentra la máquina actualmente. */
    public PaymentState state() {
        return current;
    }

    /** {@code true} si el pago no puede transicionar. */
    public boolean isTerminal() {
        return allowedEvents().isEmpty();
    }

    /** Eventos aceptados desde el estado actual. */
    public Set<PaymentEvent> allowedEvents() {
        return Collections.unmodifiableSet(
                TRANSITIONS.getOrDefault(current, Map.of()).keySet());
    }

    /** Indica si {@code event} sería aceptado desde el estado actual. */
    public boolean can(PaymentEvent event) {
        return TRANSITIONS.getOrDefault(current, Map.of()).containsKey(event);
    }

    /** Estado al que llevaría {@code event}, o vacío si no está permitido. */
    public Optional<PaymentState> peek(PaymentEvent event) {
        return Optional.ofNullable(TRANSITIONS.getOrDefault(current, Map.of()).get(event));
    }

    /**
     * Aplica {@code event} y avanza al siguiente estado.
     *
     * @return el nuevo estado
     * @throws IllegalTransitionException si el evento no está permitido en este estado
     */
    public PaymentState dispatch(PaymentEvent event) {
        PaymentState next = TRANSITIONS.getOrDefault(current, Map.of()).get(event);
        if (next == null) throw new IllegalTransitionException(current, event);
        current = next;
        return current;
    }
}
